### ISSUE TITLE
Is it possible to use as one query instead of stream?

### ISSUE THREAD

**Issue by felipecesar42** [DarshanGowda0/GeoFlutterFire#27]
I need to use the geolocation query just once in a while instead of keeping a stream open. Is there a way to do it in the current implementation?

If anyone had an idea how to implement this in the plugin I can help with the implementation/PR

**Comment by DarshanGowda0** [comment]
[This](https://github.com/DarshanGowda0/GeoFlutterFire/blob/ca14deb1cf899d4fed9011dbd447a8e1323e312a/lib/src/collection.dart#L145) is the actual query, and its a run on 9 geo-hashes - 1 center, and 8 surroundings which is then used to create streams as shown [here](https://github.com/DarshanGowda0/GeoFlutterFire/blob/ca14deb1cf899d4fed9011dbd447a8e1323e312a/lib/src/collection.dart#L90). So, **there's no single query that can be exposed**! The streams from all 9 queries are merged and a further filtered on distance. Hope this helps

**Comment by felipecesar42** [comment]
Thank you @DarshanGowda0. I need to query just once in a while for a limited number of results.  I'm developing a newsfeed of posts that has a distance filter. So soon I gonna try to get the results just once instead of keep listening all the time for the answers.

Your reply gave me some ideas on how to do it :).

**Comment by DarshanGowda0** [comment]
Happy to help 😄 I guess we'll need a getDocuments() with Futures implementation as well, along with streams?

**Comment by felipecesar42** [comment]
Yes. Have something like `getDocuments()` would be very helpful. I'm finishing other features, and then I'll come back to this need :)

**Comment by mikeyyg96** [comment]
@felipecesar42 Did you manage to get only 1 result to pop up? I also don't want it to listen all the time.

**Comment by felipecesar42** [comment]
Not yet. I'm finishing other things before being able to focus on this feature. But if you decide to start I can try to help :)

**Comment by henningvs** [comment]
> [This](https://github.com/DarshanGowda0/GeoFlutterFire/blob/ca14deb1cf899d4fed9011dbd447a8e1323e312a/lib/src/collection.dart#L145) is the actual query, and its a run on 9 geo-hashes - 1 center, and 8 surroundings which is then used to create streams as shown [here](https://github.com/DarshanGowda0/GeoFlutterFire/blob/ca14deb1cf899d4fed9011dbd447a8e1323e312a/lib/src/collection.dart#L90). So, **there's no single query that can be exposed**! The streams from all 9 queries are merged and a further filtered on distance. Hope this helps

I was wondering when looking at the method if that actually still works if you get near the poles of the earth - a location where the tiles tend to get very small? I would think that `int precision = Util.setPrecision(radius);` in collection.dart should not only take the radius into account but also the location of the query.

**Comment by awhitford** [comment]
I had a query like this:
```dart
    return Firestore.instance
        .collection('places')
        .getDocuments()
        .then((data) => data.documents
            .map((doc) => Place(
                  id: doc["id"],
                  title: doc["title"],
                ))
            .toList());
```

I changed it to use `within` like:
```dart
    return _geo
        .collection(collectionRef: Firestore.instance.collection('places'))
        .within(
            center: center,
            radius: 100,
            field: 'geoLocation',
            strictMode: true)
        .first
        .then((documents) => documents
            .map((doc) => Place(
                id: doc["id"],
                title: doc["title"],
                ))
            .toList());
```

Note that I used `first` to transform the `Stream` into a `Future`.

**Comment by gnalFF** [comment]
I am having a similar issue, however the reasoning to have a one time shot query is that the underlying stream is not canceled, when using @awhitford way of converting it into a future
The collection I query is protected by security rules, which leads then to exceptions on the stream, when logging out:

W/Firestore(15056): (19.0.0) [Firestore]: Listen for Query(rides order by start_location.location.geohash, __name__) failed: Status{code=PERMISSION_DENIED, description=Missing or insufficient permissions., cause=null}
I/System.out(15056): com.google.firebase.firestore.FirebaseFirestoreException: PERMISSION_DENIED: Missing or insufficient permissions.

Maybe I oversee something, looking forward for feedback

**Comment by felipecesar42** [comment]
Hi,

I just came back to this feature. Looks like it's impossible to convert to a Query. The best thing we can do is limit it artificially: create some kind of listener that keeps counting the number of documents retrieved and after reaching the threshold it cancels the subscription.

I searched for other implementations of this kind of query in other platforms. Looks like even in the web platforms (that have much more users than Flutter) they couldn't find a way to get this just a Query instead of Streams. Hopefully someday Firebase will improve their queries features, it's very limited so far and there's not a simple way to do complex queries even if the query isn't related to geolocation.

If someone finds another library in other platforms that solved this problem I would be happy to look their code and bring to this package.

@DarshanGowda0 do you know if I cancel the stream received from the `within` function if it will cancel all the substreams associated with it?

**Comment by klaszlo8207** [comment]
Any news on this? This feature is missing!

**Comment by theochampion** [comment]
Agreed, this would be very useful

**Comment by MartinJLee** [comment]
This seems work. 
```dart
    List<Place> ret = [];
    await geo
        .collection(collectionRef: queryRef)
        .within(center: center, radius: radius, field: 'location')
        .first
        .then((docs) {
      Future.forEach(docs, (DocumentSnapshot doc) {
        // parse place
       ret.add(place)
      }).whenComplete(() {
        return;
      });
    });
    return ret;
```
I'm not sure about the race condition this may cause. e.g. not all 9 queries available for the first steam.
It would be nice if someone can confirm this.

**Comment by lizbrownwood** [comment]
@MartinJLee , how would you get distance like this?

**Comment by MartinJLee** [comment]
> @MartinJLee , how would you get distance like this?

I'm not sure what distance you are after. Share your query code here so others can help you.

**Comment by webertim** [comment]
@lizbrownwood Do you mean something like this #87 ?
There is a pull request to, which has to be accepted ( #88 ).

**Comment by rinati** [comment]
any update on that?
I  need a Future too. Listening all the time will lead to unnecessary bills from Firestore.

**Comment by yulkin2002** [comment]
same here, would like to avoid unnecessary bills from Firestore for listening continuously

**Comment by akcadev** [comment]
The same problem continues..

**Comment by Beautylivery** [comment]
bump

**Comment by awaik** [comment]
Hey guys,
sorry for such a delay.
The author is very busy, from now on we help him to maintain the package.

@webertim I'll check your pull request. I think there will be some fixes because of such a delay with merging.
Would you like to fix them or should we do it?

Just for your info - now we have two  working branches - master and 3x_null_safety.

**Comment by giorgio79** [comment]
How about using a [StreamController ](https://api.dart.dev/stable/2.19.2/dart-async/StreamController-class.html)like so https://stackoverflow.com/questions/60663818/flutter-how-to-refresh-streambuilder
